### PR TITLE
Specification for undelegated test, issue 495

### PR DESCRIPTION
**Pull request by matsduf**
Specification for un-delegated test.

### CODE REVIEW

**Comment by sandoche2k** [comment]
@sandoche2k To review.

**Comment by mattias-p** [comment]
This introduces a new directory. The new directory should have a README file describing its purpose. The parent directory README should probably be updated too so that it links to its child directory READMEs in order to make the structure more navigable.

**Comment by sandoche2k** [comment]
1. All information about the directory should be in the parent README.md
2. Not convinced about including this test under the "Methods" directory, rather it should be part of the test specifications.

File: docs/specifications/methods/undelegated-test.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Undelegated tests
+
```
Comment (sandoche2k) [inline docs/specifications/methods/undelegated-test.md]: ## Undelegate00: Routines to be performed when testing an undelegated domain

File: docs/specifications/methods/undelegated-test.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+## Background
+
+The purpose of an undelegated test is the test a future delegation before it is
+put in production.
+
```
Comment (sandoche2k) [inline docs/specifications/methods/undelegated-test.md]: ### Objective

The purpose of an undelegated test is the test a future delegation before it is put in production. Hence, while testing an undelegated test, when extra information (such as name servers, IP addresses) are provided as input

File: docs/specifications/methods/undelegated-test.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+## Specification
+
+An undelegated test should mimic the a delegated test. The information for the 
+test must be provided when starting the test.
+
```
Comment (sandoche2k) [inline docs/specifications/methods/undelegated-test.md]: ### Inputs
The domain name to be tested.

File: docs/specifications/methods/undelegated-test.md (hunk 1 above)
Comment (matsduf) [reply docs/specifications/methods/undelegated-test.md]: The motivation for having a separate directory is that the specifications under "test" are connected to the test cases, whereas this specification defines how the environment is set for undelegated tests.

File: docs/specifications/README.md [hunk 4]
```diff
@@ -3,9 +3,12 @@
 
 ### Purpose of this directory
 
-This contents of this directory includes detailed specifications for all the
-DNS tests validated by the "zonemaster-engine".
+This contents of this directory includes detailed technical specifications for all the
```
Comment (mattias-p) [inline docs/specifications/README.md]: "This contents of this directory includes" -> "This directory contains the"?

File: docs/specifications/methods/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+![Zonemaster](https://github.com/dotse/zonemaster/blob/master/docs/images/zonemaster_logo_black.png)
+==========
+
+### Purpose of this directory
+
+This contents of this directory includes contains other specification, currently only
```
Comment (mattias-p) [inline docs/specifications/methods/README.md]: "This contents of this directory includes" -> "This directory"?
"specification" -> "specifications"?

File: docs/specifications/methods/undelegated-test.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# Undelegated tests
+
+## Objective
+
+The purpose of an undelegated test is the test a future delegation before it is
+put in production.
+
+The purpose of an undelegated test is to test a possible future delegation
```
Comment (mattias-p) [inline docs/specifications/methods/undelegated-test.md]: This sentence is almost identical to the previous one. Remove one of them.

File: docs/specifications/methods/undelegated-test.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+An undelegated test should mimic the a delegated test. The information for the 
+test must be provided when starting the test.
+
+An undelegated test can be run on a already delegated domain, but then
+Zonemaster must disregard the specific information for that domain found
+in public DNS.
+
+Information that is part of the delegation must be provided with the initiation of
+the test. That includes the following information:
```
Comment (mattias-p) [inline docs/specifications/methods/undelegated-test.md]: "Includes" is inappropriate here. It doesn't rule out the possibility that other information might also be required. The wording should make clear that the following list is exhaustive.

File: docs/specifications/methods/undelegated-test.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+* Glue records. Name server names (RDATA of the NS records) that belong to the delegated
+zone or below must be provided. They must not be looked up before the test starts.
+
+* DS records. If no DS records are provided, then it is assumed that there are no DS
+records for the zone. They must not be looked up.
+
+* Name server addresses. Addresses of name server names that belong to the delegated
+zone is covered under "glue redords" above. Addresses of other name servers may be
+provided, of else they are looked up. If at least one address (IPv4 or IPv6) is provided
```
Comment (mattias-p) [inline docs/specifications/methods/undelegated-test.md]: "of else" -> "or else"

File: docs/specifications/methods/undelegated-test.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+for a name server, then no further lookup shall be done for that name during the 
+test (neither IPv4 nor IPv6).
+
+The complete set of "NS records", the "Glue records" and the "DS records" is considered
```
Comment (mattias-p) [inline docs/specifications/methods/undelegated-test.md]: Maybe extend this paragraph to say something about the role of "Name server addresses" as well?

File: docs/specifications/README.md [hunk 10]
```diff
@@ -3,9 +3,12 @@
...
+DNS tests performed by Zonemaster.
 
-All the test implementation is based on these specifications. The documentation
-in this directory will be useful for developers, DNS experts etc.
+The [tests](tests/README.md) directory contains the test cases that the implementation of Zonemaster
+Engine is based on. The documentation in this directory will be useful for developers, 
+DNS experts etc.
 
+The [methods](methods/README.md) directory contains other specification, currently only specification for
```
Comment (mattias-p) [inline docs/specifications/README.md]: We should apply better naming here. The name "method" has already been given to [another concept](https://github.com/dotse/zonemaster/blob/master/docs/specifications/tests/Methods.md). I suggest calling the directory "test-types" instead, and maybe rename the "tests" directory to "test-cases".

File: docs/specifications/README.md (hunk 4 above)
Comment (matsduf) [reply docs/specifications/README.md]: Ok.

File: docs/specifications/README.md (hunk 10 above)
Comment (matsduf) [reply docs/specifications/README.md]: I change to "test-types". Changing from "tests" to "test-cases" affects many files, so I'll do that in a different change.

File: docs/specifications/methods/README.md (hunk 5 above)
Comment (matsduf) [reply docs/specifications/methods/README.md]: Will fix.

File: docs/specifications/methods/undelegated-test.md (hunk 6 above)
Comment (matsduf) [reply docs/specifications/methods/undelegated-test.md]: Will fix.

File: docs/specifications/test-types/README.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+![Zonemaster](https://github.com/dotse/zonemaster/blob/master/docs/images/zonemaster_logo_black.png)
+==========
+
+### Purpose of this directory
+
+This includes contains **Test Type** specifications, currently only
```
Comment (mattias-p) [inline docs/specifications/test-types/README.md]: "includes contains" -> "directory contains"

**Comment by sandoche2k** [comment]
@matsduf Ok for me.

**Comment by matsduf** [comment]
This PR has been created against the master branch. Close this and open a new PR against the develop branch.

**Pull request closed without merging**